text/0000-improved-ember-registry-apis.md

```diff
@@ -0,0 +1,719 @@
+
+### Do nothing
+
+Leaving the API as it is remains an option, with both the upsides and downsides of the _status quo_.
+
+## Unresolved questions
+
+- Is `@ember/application` the appropriate home for the new exports? Or should they liver somewhere else, such as `@ember/resolver` or `@ember/application/resolver`?
+
+- What is the right name for the parameter and interface for the lookup? `identifier` is a reasonable choice, but does overlap with the notion of identifiers from Ember Data. It is also *quite* generic. We could use any of a number of longer but perhaps clearer names:
+    - `RegistrationIdentifier`
+    - `RegistryIdentifier`
+    - `FactoryIdentifier`
+    - `RegistryId`
+
+- Similarly, is `FactoryTypeIdentifier` the correct name? Alternatives might be:
+    - `FactoryRegistrationIdentifier`
+    - `FactoryTypeRegistrationIdentifier`
+    - `FactoryRegistryIdentifier`
+    - `FactoryTypeRegistryIdentifier`
+    - `FactoryIdentifier` (mutually exclusive with using it as a replacement for `Identifier`
+    - `FactoryRegistryId`
+    - `FactoryTypeRegistryId`
+
+- Are `identifier` and `type` the right names for the arguments in the new design?
+
+## Appendix: TypeScript
+
+This design is motivated in part by a desire for the API to better support type-safe TypeScript usage. This appendix explains the motivation and design as it impacts the declarations maintained by the Typed Ember team. The big idea here is to make registry lookups type-safe, including support for autocomplete and refactoring, while not making the boilerplate situation *worse* (eliminating it entirely is unfortunately not yet possible).
+
+### TypeScript motivation
+
+The type definitions for Ember currently make use of a pattern we call a “type registry”: a mapping of *strings* to *types*. This allows the types to safely represent string-based lookups like `story.findRecord('some-model')` or `service('session')`.
+
+For example, to make service lookups work with the classic API, users had to write something like this:
+
+```ts
+import Service from '@ember/service';
+
+export default class Session extends Service {
+  login(email: string, password: string) {
+    // ...
+  }
+  
+  logout() {
+    // ...
+  }
+}
+
+declare module '@ember/service' {
+  interface Registry {
+    session: Session;
+  }
+}
+```
+
+Then, users could write fairly idiomatic Ember and have the types resolve correctly:
+
+```ts
+import Component from '@ember/component';
+import { inject as service } from '@ember/service';
+
+export default Component.extend({
+  session: service('session'), // type: `Session`
+})
+```
+
+The Typed Ember blueprints generate that boilerplate for users, which (hopefully) minimizes the pain of working with these definitions. However, to make this work for the registry APIs, users (and therefore the blueprints) would have to add *another* declaration to the bottom of the file:
```

Comment:
This is _well_ outside the scope of this RFC, but one of the things I'd love to explore (likely within the realm of that `@typeset` project we've discussed) is automating the maintenance of these type registries instead of relying on boilerplate merged from the source modules.

Reply:
I had that point very much in mind myself as I was writing this. If we can programmatically generate/maintain those, that would make users' lives *much* easier. I experimented with that a bit in my previous gig, but never got to something I like.